lib/plug_caisson.ex

```diff
@@ -42,27 +42,51 @@ defmodule PlugCaisson do
   @spec read_body(Plug.Conn.t()) ::
           {:ok, binary(), Plug.Conn.t()} | {:more, binary(), Plug.Conn.t()} | {:error, term()}
   def read_body(conn, opts \\ []) do
-    content_encoding = Plug.Conn.get_req_header(conn, "content-encoding")
-    types = opts[:algorithms] || @default
-
-    with {:ok, body, conn} <- Plug.Conn.read_body(conn, opts) do
-      case try_decompress(body, content_encoding, types) do
-        {:ok, data} -> {:ok, data, conn}
-        {:more, data} -> {:more, data, conn}
-        {:error, _} = error -> error
+    with {:ok, decoder, conn} <- fetch_decompressor(conn, opts[:algorithms] || @default) do
+      case Plug.Conn.read_body(conn, opts) do
+        {type, body, conn} when type in [:ok, :more] ->
+          case try_decompress(body, decoder) do
+            {:error, _} = error -> error
+            {:ok, data} when type == :ok -> {:ok, data, conn}
+            {_, data} -> {:more, data, conn}
+          end
+
+        {:error, _} = error ->
+          error
```

Comment:
That do not have the same behaviour as in case of `:more` clause we always return `:more` (as the underlying decompression method may return `:ok` in case of successful partial decompression).

Reply:
The guard clause can use previously bound variables as well, so referencing type `type` instead so that the correct atom is returned will work.
Main issue is readability due to the 3 level nesting without much benefit of using case/with since here are only 1 main matched clause for each level.